### ISSUE TITLE
Full Coverage Integration Tests

### ISSUE THREAD

**Issue by eahefnawy** [serverless/integration-test-suite#8]
Our current integration test suite is pretty basic and needs enhancements. Here's  a list of additional tests that we should add that should cover each angle of framework usage.

#### S3
* single s3 event should create bucket, subscribe lambda function, trigger S3 event and make sure lambda is invoked.
* multiple s3 events with different buckets should create all buckets, subscribe corresponding lambda functions, trigger S3 events and make sure lambdas are invoked.
* multiple s3 events with same bucket name should create a single bucket, subscribe corresponding lambda functions, trigger S3 events and make sure lambda is invoked

#### Schedule
* single schedule event should invoke function at intended interval
* multiple schedule events for the same lambda should invoke function at all intended intervals

#### SNS
* single sns event should create topic, subscribe function, publish to topic and make sure lambda is invoked
* multiple sns events with different topics should create all topics, subscribe corresponding lambda functions, publish to topic and make sure lambdas are invoked

#### API Gateway
* simple REST API: multiple methods for the same path
* enable cors and make sure correct headers/origins/methods are set
* enable authorizer function, make a request, make sure authorizer is invoked and make sure the request succeeds
* add api keys, make unauthenticated request and make sure it fails then make authenticated request and make sure it succeeds

#### Functions
* test packaging: make sure specified packaging configuration is working
* invoke function with test data
* test deep handler files
* test function with vpc config
* test function with custom IAM roles

#### Resources
* Make sure custom resources are created
* Make sure custom outputs are created
* Make sure our logical IDs system is not violated

#### Plugins
* Make sure we can load custom plugins, and call its commands

**Comment by pmuens** [comment]
Great list! Thanks for that @eahefnawy 

What do you think about the `iamRoleStatements` on can setup through the `provider` property. Should we also make sure that those are created and attached correctly or are those unit tests in the main repo sufficient?

**Comment by eahefnawy** [comment]
@pmuens yeah imo we can include those in the function tests since they're about giving the role the function is using more access
